pipeline/lib/metric.ml

```diff
@@ -0,0 +1,100 @@
+module Github = Current_github
+
+module MetricKey = struct
+  type t = string * string * string
+
+  let compare a b = compare a b
+end
+
+module BenchmarksData = Map.Make (MetricKey)
+
+let parse_benchmark_data ~data =
+  Array.fold_left
+    (fun acc datum ->
+      match datum with
+      | _commit, benchmark_name, test_name, metrics ->
+          metrics
+          |> List.fold_left
+               (fun acc ({ name; value; _ } : Current_bench_json.Latest.metric) ->
+                 let key = (benchmark_name, test_name, name) in
+                 BenchmarksData.add key value acc)
+               acc)
+    BenchmarksData.empty data
+
+let sum_float_list xs = List.fold_left (fun acc x -> acc +. x) 0. xs
+
+let mean_float_list xs =
+  let n = List.length xs in
+  match n with 0 -> 0. | _ -> sum_float_list xs /. Float.of_int n
+
+let calc_diff v cv = if cv <> 0. then (v -. cv) /. cv *. 100. else 0.
+
+let find_avg vs =
+  vs
+  |> List.filter_map (function
+       | key, value when key = "avg" -> Some value
+       | _ -> None)
+  |> function
+  | [ avg ] -> avg
+  | _ -> 0.
+
+let change_threshold = 0.01
+
+let find_changed_metrics ~metrics ~compare_metrics =
+  BenchmarksData.merge
+    (fun _ (value : Current_bench_json.Latest.value option)
+         (compare_value : Current_bench_json.Latest.value option) ->
+      let delta =
+        match (value, compare_value) with
+        | Some value, Some compare_value -> (
+            match (value, compare_value) with
+            | Float v, Float cv -> calc_diff v cv
+            | Floats vs, Floats cvs ->
+                let v = mean_float_list vs in
+                let cv = mean_float_list cvs in
+                calc_diff v cv
+            | Assoc vs, Assoc cvs -> calc_diff (find_avg vs) (find_avg cvs)
+            | _ -> 0.)
+        | _ -> 0.
+      in
+      match delta with
+      | _ when Float.abs delta > change_threshold -> Some delta
+      | _ -> None)
```

Comment:
To clarify some earlier comment, I think I would go with something like this:
```ocaml
BenchmarksData.merge
  (fun _ value compare_value -> match value, compare_value with
    | Some value, Some compare_value ->
      let delta = calc_diff (mean value) (mean compare_value) in
      if Float.abs delta > change_threshold then Some delta else None
    | _ -> None)
```
We might want to report the "missing benchmarks" (with `| Some _, None -> ...`, but no delta condition this time)

Reply:
This makes sense. 

I don't know if we should also flag the case where `cv = 0`; `calc_diff` currently returns `0.` which would drop the metric when comparing against the threshold.

Reply:
Yeah, zero will always be a bit weird... Perhaps the formula could be "is `cv` within ±10% of `v`" ? (to avoid the division)

Reply:
Yeah, that seems like a reasonable way to handle this! :+1:

Reply:
I changed `calc_diff` to divide with whichever value is non-zero (`a` or `b`). It might be nice to display the %age change in the GitHub comment, so I took this approach.